internal/wasm/sdk/test/e2e/external_test.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2020 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+// +build wasm_sdk_e2e
+
+package e2e
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/internal/wasm/sdk/opa"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/test/cases"
+	"github.com/open-policy-agent/opa/types"
+	"github.com/open-policy-agent/opa/util"
+)
+
+var opaRootDir = "../../../../../"
+var caseDir = flag.String("case-dir", filepath.Join(opaRootDir, "test/cases/testdata"), "set directory to load test cases from")
+var exceptionsFile = flag.String("exceptions", "./exceptions.yaml", "set file to load a list of test names to exclude")
+
+var exceptions map[string]string
+
+func TestMain(m *testing.M) {
+	exceptions = map[string]string{}
+
+	bs, err := ioutil.ReadFile(*exceptionsFile)
+	if err != nil {
+		fmt.Println("Unable to load exceptions file: " + err.Error())
+		os.Exit(1)
+	}
+	err = util.Unmarshal(bs, &exceptions)
+	if err != nil {
+		fmt.Println("Unable to parse exceptions file: " + err.Error())
+		os.Exit(1)
+	}
+
+	addTestSleepBuiltin()
+
+	os.Exit(m.Run())
+}
+
+func TestWasmE2E(t *testing.T) {
+
+	ctx := context.Background()
+
+	for _, tc := range cases.MustLoad(*caseDir).Sorted().Cases {
+		name := fmt.Sprintf("%s/%s", strings.TrimPrefix(tc.Filename, opaRootDir), tc.Note)
+		t.Run(name, func(t *testing.T) {
+
+			if shouldSkip(t, tc) {
+				t.SkipNow()
+			}
+
+			opts := []func(*rego.Rego){
+				rego.Query(tc.Query),
+			}
+			for i := range tc.Modules {
+				opts = append(opts, rego.Module(fmt.Sprintf("module-%d.rego", i), tc.Modules[i]))
+			}
+			cr, err := rego.New(opts...).Compile(ctx)
+			if err != nil {
+				t.Fatal(err)
+			}
+			o := opa.New().WithPolicyBytes(cr.Bytes)
+			if tc.Data != nil {
+				o = o.WithDataJSON(tc.Data)
+			}
+			o, err = o.Init()
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			var input *interface{}
+
+			if tc.InputTerm != nil {
+				var x interface{} = ast.MustParseTerm(*tc.InputTerm)
+				input = &x
+			} else if tc.Input != nil {
+				input = tc.Input
+			}
+
+			result, err := o.Eval(ctx, opa.EvalOpts{Input: input})
+			assert(t, tc, result, err)
+		})
+	}
+}
+
+func shouldSkip(t *testing.T, tc cases.TestCase) bool {
+	if tc.WantError != nil || tc.WantErrorCode != nil {
+		t.Log("Skipping test case: Error validation not supported")
+		return true
+	}
+
+	if reason, ok := exceptions[tc.Note]; ok {
+		t.Log("Skipping test case: " + reason)
+		return true
+	}
+
+	return false
+}
+
+func assert(t *testing.T, tc cases.TestCase, result *opa.Result, err error) {
+	t.Helper()
+	if tc.WantDefined != nil {
+		if err != nil {
+			t.Fatal("unexpected error:", err)
+		} else {
+			assertDefined(t, defined(*tc.WantDefined), result)
+		}
```

Comment:
`t.Fatal` stops execution, so we could get rid of some nesting here (and below)
```suggestion
		if err != nil {
			t.Fatal("unexpected error:", err)
		}
		assertDefined(t, defined(*tc.WantDefined), result)
```

Reply:
Fixed